components/brave_shields/browser/tracking_protection_service.cc

```diff
@@ -37,6 +51,145 @@ TrackingProtectionService::~TrackingProtectionService() {
   tracking_protection_client_.reset();
 }
 
+#if BUILDFLAG(BRAVE_STP_ENABLED)
+TrackingProtectionService::RenderFrameIdKey::RenderFrameIdKey()
+    : render_process_id(content::ChildProcessHost::kInvalidUniqueID),
+      frame_routing_id(MSG_ROUTING_NONE) {}
+
+TrackingProtectionService::RenderFrameIdKey::RenderFrameIdKey(
+    int render_process_id,
+    int frame_routing_id)
+    : render_process_id(render_process_id),
+      frame_routing_id(frame_routing_id) {}
+
+bool TrackingProtectionService::RenderFrameIdKey::operator<(
+    const RenderFrameIdKey& other) const {
+  return std::tie(render_process_id, frame_routing_id) <
+         std::tie(other.render_process_id, other.frame_routing_id);
+}
+
+bool TrackingProtectionService::RenderFrameIdKey::operator==(
+    const RenderFrameIdKey& other) const {
+  return render_process_id == other.render_process_id &&
+         frame_routing_id == other.frame_routing_id;
+}
+
+void TrackingProtectionService::SetStartingSiteForRenderFrame(
+                                                         GURL starting_site,
+                                                         int render_process_id,
+                                                         int render_frame_id) {
+  base::AutoLock lock(frame_starting_site_map_lock_);
+  const RenderFrameIdKey key(render_process_id, render_frame_id);
+  render_frame_key_to_starting_site_url[key] = starting_site;
+}
+
+GURL TrackingProtectionService::GetStartingSiteForRenderFrame(
+  int render_process_id, int render_frame_id) {
+  base::AutoLock lock(frame_starting_site_map_lock_);
+  const RenderFrameIdKey key(render_process_id, render_frame_id);
+  auto iter = render_frame_key_to_starting_site_url.find(key);
+  if (iter != render_frame_key_to_starting_site_url.end()) {
+    return iter->second;
+  }
+  return GURL();
+}
+
+void TrackingProtectionService::ModifyRenderFrameKey(int old_render_process_id,
+                                                     int old_render_frame_id,
+                                                     int new_render_process_id,
+                                                     int new_render_frame_id) {
+  base::AutoLock lock(frame_starting_site_map_lock_);
+  const RenderFrameIdKey old_key(old_render_process_id, old_render_frame_id);
+  auto iter =
+      render_frame_key_to_starting_site_url.find(old_key);
+  if (iter != render_frame_key_to_starting_site_url.end()) {
+    const RenderFrameIdKey new_key(new_render_process_id, new_render_frame_id);
+    render_frame_key_to_starting_site_url.insert(
+        std::pair<RenderFrameIdKey, GURL>(new_key, iter->second));
+    render_frame_key_to_starting_site_url.erase(old_key);
+  }
```

Comment:
`else { NOTREACHED(); }` ? Is it a possible situation?

Reply:
I had this code initially, but it looks like `DidStartNavigation` and `ReadyToCommitNavigation` are not invoked when preferences are opened from menu `Brave -> Preferences` or `Brave ->ClearBrowsingData`. However, `RenderFrameHostChanged` is invoked when the page is rendered.

Reply:
hmm, ok, good observation!